File: csharp/parsing-log-files/ParsingLogFiles.cs
using System.Text.RegularExpressions;

public class LogParser
{
    public bool IsValidLine(string text)
    {
        Regex re = new(@"^\[(TRC|DBG|INF|WRN|ERR|FTL)\].*$");
        return re.IsMatch(text);        
    }

    public string[] SplitLogLine(string text)
    {
        Regex re = new(@"\<(\^|\*|=|-)+\>");

        List<string> results = re.Split(text).ToList<String>();
        return results.Where((value, count) => count % 2 == 0).ToArray<string>();
    }

    public int CountQuotedPasswords(string lines)
    {
        Regex re = new(@""".*password.*""", RegexOptions.IgnoreCase);
        MatchCollection mc = re.Matches(lines);
        return mc.Count();
    }

    public string RemoveEndOfLineText(string line)
    {
        Regex re = new(@"end-of-line\d+");
        return re.Replace(line, "");
    }

    public string[] ListLinesWithPasswords(string[] lines)
    {
        Regex re = new(@"password\w+", RegexOptions.IgnoreCase);
        List<string> results = new();
        foreach (string line in lines)
        {
            Match m = re.Match(line);
            if (m.Captures.Count > 0)
            {
                results.Add($"{m.Value}: {line}");
            }
            else
            {
                results.Add($"--------: {line}");
            }
        }
        return results.ToArray<string>();
    }
}

File: go/parsing-log-files/parsing_log_files.go
package parsinglogfiles

import (
	"fmt"
	"regexp"
)

func IsValidLine(text string) bool {
	re := regexp.MustCompile(`^\[(TRC|DBG|INF|WRN|ERR|FTL)\]`)
	return re.MatchString(text)
}

func SplitLogLine(text string) []string {
	re := regexp.MustCompile(`\<[~\*-=]*\>`)
	return re.Split(text, -1)
}

func CountQuotedPasswords(lines []string) int {
	re := regexp.MustCompile(`(?i)\".*password.*\"|\'.*password.*\'`)
	passwordCount := 0
	for _, line := range lines {
		if re.MatchString(line) {
			passwordCount++
		}
	}
	return passwordCount
}

func RemoveEndOfLineText(text string) string {
	re := regexp.MustCompile(`end-of-line\d+`)
	return re.ReplaceAllString(text, "")
}

func TagWithUserName(lines []string) []string {
	re := regexp.MustCompile(`User +([a-zA-Z][a-zA-Z0-9]*) `)
	var result []string
	for _, line := range lines {
		find_result := re.FindStringSubmatch(line)
		if len(find_result) > 0 {
			result = append(result, fmt.Sprintf("[USR] %s %s", find_result[1], line))
		} else {
			result = append(result, line)
		}
	}
	return result
}
